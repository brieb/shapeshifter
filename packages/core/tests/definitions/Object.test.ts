import { options, truthyValues } from '../mocks';
import ObjectDefinition from '../../src/definitions/Object';
import NumberDefinition from '../../src/definitions/Number';
import StringDefinition from '../../src/definitions/String';

describe('definitions/Object', () => {
  it('errors if `valueType` is missing', () => {
    expect(() => new ObjectDefinition(options, 'foo', {})).toThrow(
      'Invalid ObjectDefinition field "valueType". Field is required and must be defined.',
    );
  });

  it('errors if `valueType` is empty', () => {
    expect(
      () => new ObjectDefinition(options, 'foo', { valueType: '' }),
    ).toThrowErrorMatchingSnapshot();
  });

  it('errors if `keyType` is empty', () => {
    expect(
      () => new ObjectDefinition(options, 'foo', { keyType: '', valueType: 'string' }),
    ).toThrow('Invalid ObjectDefinition field "keyType". String cannot be empty.');
  });

  it('errors if `keyType` is not a string', () => {
    truthyValues
      .filter((value) => typeof value !== 'string')
      .forEach((value) => {
        expect(
          // @ts-expect-error
          () => new ObjectDefinition(options, 'foo', { keyType: value, valueType: 'string' }),
        ).toThrow('Invalid ObjectDefinition field "keyType". Must be a string.');
      });
  });

  it('creates a `Definition` for the `valueType`', () => {
    const def = new ObjectDefinition(options, 'foo', { valueType: 'number' });

    expect(def.valueType).toBeInstanceOf(NumberDefinition);
  });

  it('creates a `Definition` for the `keyType`', () => {
    const def = new ObjectDefinition(options, 'foo', { valueType: 'number', keyType: 'string' });

    expect(def.keyType).toBeInstanceOf(StringDefinition);
  });
});
