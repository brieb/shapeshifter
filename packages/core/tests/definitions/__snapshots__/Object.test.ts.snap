// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`definitions/Object errors if \`valueType\` is empty 1`] = `
"Invalid ObjectDefinition field \\"valueType\\". Type must be one of: string, shape. Received string with the following invalidations:
 - String cannot be empty."
`;
